pkg/codegen/python/gen.go

```diff
@@ -1348,7 +1348,8 @@ func (mod *modContext) genResource(res *schema.Resource) (string, error) {
 		// Emit logic to configure schema.ObjectType args that have defaults and deprecation messages.
 		if objType, ok := unwrapped.(*schema.ObjectType); ok {
 			expectedObjType := mod.objectClassRef(objType, true)
-			fmt.Fprintf(w, "            if not isinstance(%s, %s):\n", InitParamName(prop.Name), expectedObjType)
+			fmt.Fprintf(w, "            if %[1]s is not None and not isinstance(%[1]s, %s):\n",
+				InitParamName(prop.Name), expectedObjType)
```

Comment:
I understand the original implementation of this was `if isinstance(%s, dict)`, which would have avoided this error, but the code review feedback was to flip the check around to check for the args class instead, to be more permissive about the types of dictionaries that could be supported. https://github.com/pulumi/pulumi/pull/13825#discussion_r1317218561.

I still wonder if we should switch this back to checking for a dictionary, but rather than checking for `dict` specifically, check if it is an instance of `Mapping` (`if isinstance(%s, Mapping)`), which would be permissive in the types of dictionaries it accepts.

I've chosen not to do that for this fix, but wanted to mention it.